pkg/agent/agent.go

```diff
@@ -1132,23 +1132,19 @@ func getRoundInfo(bridgeClient ovsconfig.OVSBridgeClient) types.RoundInfo {
 	return roundInfo
 }
 
-func (i *Initializer) getNodeMTU(transportInterface *net.Interface) (int, error) {
+func (i *Initializer) getInterfaceMTU(transportInterface *net.Interface) (int, error) {
 	if i.mtu != 0 {
 		return i.mtu, nil
 	}
 	mtu := transportInterface.MTU
-	// Make sure mtu is set on the interface.
+	// Make sure MTU is set on the interface.
 	if mtu <= 0 {
 		return 0, fmt.Errorf("Failed to fetch Node MTU : %v", mtu)
 	}
+
+	mtu -= i.networkConfig.MTUDeduction
+
 	if i.networkConfig.TrafficEncapMode.SupportsEncap() {
-		if i.networkConfig.TunnelType == ovsconfig.VXLANTunnel {
-			mtu -= config.VXLANOverhead
-		} else if i.networkConfig.TunnelType == ovsconfig.GeneveTunnel {
-			mtu -= config.GeneveOverhead
-		} else if i.networkConfig.TunnelType == ovsconfig.GRETunnel {
-			mtu -= config.GREOverhead
-		}
 		if i.nodeConfig.NodeIPv6Addr != nil {
```

Comment:
I wonder when we do IPv6 encap, when the Node has both IPv4 and IPv6 addresses? @wenyingd 

For Multi-cluster GW tunnel, should we keep the same behavior if possible? It is hard to understand otherwise.

Reply:
Do you mean how to encap the traffic with dual-stack setup? We are performing IPv4 over IPv4 and IPv6 over IPv6.

Reply:
@jianjuns Do you mean keep the behavior for mtu calculation or you mean something else? For now, networkpolicy only mode has different MTU calculation, for other modes, they are actually the same and reuse current method.

Reply:
@luolanzone : yes I am thinking should we share the same MTU calculation for NetworkPolicyOnly + MC Gateway and other modes. Otherwise it is quite hard to understand. It means in the dual stack case, we will include IPv6 overhead in MTUDeduction, which also applies to NetworkPolicyOnly + MC Gateway.

If we do this, we should call `networkConfig.CalculateMTUDeduction()` in `getInterfaceMTU()` and pass an `ipv6` boolean to the func.

Reply:
@jianjuns  I discussed with wenying, We feel it's more precise to get current Pod's interface and do the deduction to set the right MTU. But it's doable to share the same MTU calculation and use the new networkConfig.InterfaceMTU field for NP only interface MTU configuration directly. But the assumption would be Pod's interface MTU is the same as Node interface MTU in NP only mode(For now, it's true in AWS clusters. All are 9000). Let me know if you think the assumption is applied in NetworkPolicyOnly + MC Gateway. I will refine this part. Thanks.

Reply:
@luolanzone I am not suggesting to apply interface MTU directly to NPOnly Pods and override its original MTU. I am suggesting to calculate `NetworkConfig.MTUDeduction` with IPv6 overhead considered. Check my last comment:

> If we do this, we should call `networkConfig.CalculateMTUDeduction()` in `getInterfaceMTU()` and pass an ipv6 boolean to the func.

So, `CalculateMTUDeduction()` should consider IPv6 overhead, and `getInterfaceMTU()` just uses `NetworkConfig.MTUDeduction` to get interface MTU and no need to additionally add IPv6 overhead.

Reply:
Done, initialize the networkConfig.MTUDeduction in getInterfaceMTU().